### DISCUSSION TITLE
Image Optimization

### DISCUSSION THREAD

**Discussion by lifeiscontent** [epicweb-dev/epic-stack#80] [Q&A]
Hi, I noticed that there is an altText attribute on the image model but no usages of it in the code, how would you go about using it?

also, for things like image variants (resized images / alternative formats) how would you go about dealing with that? one idea that came to mind was to have a one to many relationship to the image model to ask for a specific variant.

e.g. if variant exists, use it, otherwise create it with `sharp` and store it for later to use with a picture element.

**Comment by kentcdodds** [comment]
For your first question, this is useful for user submitted images. You would ask them to provide the alt text and then wherever you display the image you would include the alt text on the image.

For your second question, I'll be building a special route that allows transformations similar to cloudinary (likely using `sharp`) and cache the generated images on demand. My hope is to be able to use https://github.com/ascorbic/unpic-img in some fashion.

**Reply by lifeiscontent** [reply]
@kentcdodds I'm not sure if this was the direction you wanted to go, but I built an image loader that has the same API as the next/image internal route.

if you think this is a good direction, I can open a PR. I put it under `app/routes/api+/image.tsx` in my own app, but if you like this approach, we can discuss further :)

```ts
import type { DataFunctionArgs } from '@remix-run/node'
import crypto from 'crypto'
import fs from 'fs'
import path from 'path'
import sharp from 'sharp'
import stream from 'stream'

const CACHE_DIR = path.join(process.cwd(), '.cache', 'remix-image')
const TTL = 60 * 60 * 24 // 1 day

const SUPPORTED_TYPES_CONFIG = {
	'image/jpeg': { method: 'jpeg', mimeType: 'image/jpeg' },
	'image/png': { method: 'png', mimeType: 'image/png' },
	'image/webp': { method: 'webp', mimeType: 'image/webp' },
} as const

const SUPPORTED_TYPES: (keyof typeof SUPPORTED_TYPES_CONFIG)[] = [
	'image/jpeg',
	'image/png',
	'image/webp',
]

async function getImage(params: {
	base: string
	url: string
	width: string
	quality: string
	supportedType: (typeof SUPPORTED_TYPES_CONFIG)[keyof typeof SUPPORTED_TYPES_CONFIG]
}) {
	const { url, width, quality, supportedType } = params
	const hash = crypto
		.createHash('md5')
		.update([url, width, quality, supportedType.mimeType].join('-'))
		.digest('hex')
	const filePath = path.join(CACHE_DIR, `${hash}.${supportedType.method}`)

	try {
		const stats = await fs.promises.stat(filePath)
		const now = new Date()
		if ((now.getTime() - stats.mtime.getTime()) / 1000 < TTL) {
			return fs.createReadStream(filePath)
		}
	} catch (err) {
		if ((err as NodeJS.ErrnoException).code !== 'ENOENT') {
			throw new Error(
				`Error accessing file ${filePath}${
					err instanceof Error ? `: ${err.message}` : ''
				}`,
			)
		}
	}

	const res = await fetch(url.startsWith('/') ? new URL(url, params.base) : url)

	if (!res.ok) {
		throw new Error(`Failed to fetch ${url}`)
	}

	const transformer = sharp()
		.resize(parseInt(width))
		[supportedType.method]({ quality: parseInt(quality) })

	const writeStream = fs.createWriteStream(filePath)
	await stream.promises.pipeline(
		// @ts-expect-error - this works, but TypeScript is not happy about it
		//                    I would expect stream.Readable.fromWeb to work here, but it doesn't.
		stream.Readable.from(res.body!),
		transformer,
		writeStream,
	)

	return fs.createReadStream(filePath)
}

function getCacheHeaders(url: string): Record<string, string> {
	if (url.startsWith('/build/_assets')) {
		return {
			'Cache-Control': 'public, max-age=31536000, immutable',
		}
	}

	const expires = new Date()
	expires.setSeconds(expires.getSeconds() + TTL)

	return {
		'Cache-Control': `max-age=${TTL}`,
		Expires: expires.toUTCString(),
	}
}

async function createDirectory(path: string) {
	try {
		await fs.promises.mkdir(path, { recursive: true })
	} catch (error) {
		if ((error as NodeJS.ErrnoException).code !== 'EEXIST') {
			throw error
		}
	}
}

export async function loader({ request }: DataFunctionArgs) {
	await createDirectory(CACHE_DIR)
	const url = new URL(request.url)
	const src = url.searchParams.get('url')
	const width = url.searchParams.get('w')
	const quality = url.searchParams.get('q')

	if (!src || !width || !quality) {
		return new Response(null, { status: 400, statusText: 'Bad Request' })
	}

	const requestedType = request.headers
		.get('Accept')
		?.split(',')
		.find<(typeof SUPPORTED_TYPES)[number]>(
			(t): t is (typeof SUPPORTED_TYPES)[number] => SUPPORTED_TYPES.includes(t),
		)

	const supportedType = SUPPORTED_TYPES_CONFIG[requestedType!]

	if (!supportedType) {
		return new Response(null, { status: 406, statusText: 'Not Acceptable' })
	}

	try {
		const imageStream = await getImage({
			base: url.origin,
			url: src,
			width,
			quality,
			supportedType,
		})
		return new Response(
			stream.Readable.toWeb(imageStream) as ReadableStream<Uint8Array>,
			{
				headers: {
					'Content-Type': supportedType.mimeType,
					...getCacheHeaders(src),
				},
			},
		)
	} catch (err) {
		console.error(err)
		if ((err as NodeJS.ErrnoException).code === 'ENOENT') {
			return new Response(null, {
				status: 404,
				statusText: 'Not Found',
			})
		} else {
			return new Response(null, {
				status: 500,
				statusText: 'Internal Server Error',
			})
		}
	}
}
```

**Reply by kentcdodds** [reply]
Thank you so much for sharing this @lifeiscontent!

I'll very likely use what you've built as an inspiration for what goes into this stack. One difference will be that I plan on making a special sqlite database for caching the images. I've got to figure out how to make it so I can setup LiteFS for that database to write from any region. cc @benbjohnson, isn't that possible with some trade-offs?

**Reply by lifeiscontent** [reply]
@kentcdodds thats great! Thank you for putting together epic stack, I've been taking parts and pieces of it for another project I'm building. Happy to noodle on this idea more if you want help as I'll probably end up needing whatever you're planning at some point anyway also 😀 I'm currently working on an image component as well but you mentioned using unpic so not sure if you think the code would be useful, but happy to share 😊

**Reply by benbjohnson** [reply]
@kentcdodds are you wanting to store the images themselves? We have write forwarding but it has limitations. We've actually talked about two possible features that are relevant:

1. Adding an HTTP API for queries. This would require a small JavaScript driver but it'd let you query from any node.

2. Supporting non-SQLite files. This would let you write a whole file to LiteFS and it would automatically propagate to all nodes. It would only support full files. If you replaced the file then you'd have to replace the whole thing (and not just update part of it).

Would either of those seem like good options for what you're doing?

**Reply by kentcdodds** [reply]
1. That's basically like https://turso.tech/ then? Crazy! That would definitely address this.
2. That would also solve this and be even easier. I may even switch to that for my source image storage as well. (What we're discussing is just caching of image transformations).

I think between the two of them, I'd go for option 2, but option 1 would also allow me to avoid having to do this: https://github.com/epicweb-dev/epic-stack/blob/ea4ff766b8ffd3efd3b229c8166c9f4e281263ec/app/routes/admin%2B/cache_.sqlite.tsx#L48

So... both? 😅

**Reply by benbjohnson** [reply]
Yes, number 1 is conceptually the same approach as Turso. It's not meant to be the primary access method as going over HTTP has much higher latency than in-process queries. It came up because one of the issues we have is with "background job" instances. They tend to live in the same region as the primary node but they are not the primary node themselves. In those cases, HTTP has low enough latency that it's not a big deal.

I added issues for both:

- https://github.com/superfly/litefs/issues/326
- https://github.com/superfly/litefs/issues/327

**Reply by muco-rolle** [reply]
Hi @benbjohnson 

I need this component. Could you please share the code with me?
